Controller Component Threat Model (#60)
radius-project/design-notes: architecture/2024-08-controller-component-threat-model.md

   2. **Library**: The library used to calculate the hash of the deployment configuration is the crypto library, which is one of the standard libraries of Go: [Link to library](https://pkg.go.dev/crypto@go1.23.1).
   3. **Type**: [SHA1](https://www.rfc-editor.org/rfc/rfc3174.html). Note: "SHA-1 is cryptographically broken and should not be used for secure applications." [Link to warning](https://pkg.go.dev/crypto/sha1@go1.23.1).

2. **Hashing the Secret Data**: [Link to code](https://github.com/radius-project/radius/blob/8151a96665b7f5bcd6474f5e33aff35d01adfa5a/pkg/controller/reconciler/deployment_reconciler.go#L580).

   1. **Purpose**: We hash the secret data and add it to the Pod definition to determine if the secret has changed in an update.
   2. **Library**: The library used to calculate the hash of the secret is the crypto library, which is one of the standard libraries of Go: [Link to library](https://pkg.go.dev/crypto@go1.23.1).
   3. **Type**: [SHA1](https://www.rfc-editor.org/rfc/rfc3174.html). Note: "SHA-1 is cryptographically broken and should not be used for secure applications." [Link to warning](https://pkg.go.dev/crypto/sha1@go1.23.1).

> 

Reviewer: This is a good explanation. We should mention that this is used as an optimization, not as a security protection.

#### Storage of secrets

Below you will find where and how Radius stores secrets. We create Kubernetes Secret objects and rely on Kubernetes security measures to protect these secrets.

1. **Deployment Reconciler**: Creates or updates a Kubernetes Secret for the connection values in a Deployment object. This Kubernetes Secret is deleted when the Deployment is deleted.
2. **Recipe Reconciler**: Creates a Kubernetes Secret for the Recipe object if it is defined in its spec. If there is an update to the secret object, the old one is deleted and the new one is added. When the Recipe is deleted, the Secret is also deleted.

#### Data Serialization / Formats

> We do not use custom parsers and instead rely on Kubernetes built-in parsers. Therefore, we trust Kubernetes security measures to handle data serialization and formats securely.

Reviewer: We parse resource IDs which is a custom format.

The Kubernetes API Server, which is the main interactor of the Controller component, runs in the `kube-system` namespace within the cluster.

### Key Points of Namespaces

1. **Isolation of Resources and Workloads**: Different namespaces separate and isolate resources and workloads within the Kubernetes cluster.
2. **Access Controls and Permissions**: Access controls and other permissions are implemented to manage interactions between namespaces.
3. **Separation of Concerns**: Namespaces support the separation of concerns by allowing different teams or applications to manage their resources independently, reducing the risk of configuration errors and unauthorized changes.

## Assumptions

This threat model assumes that:

1. The Radius installation is not tampered with.
2. The Kubernetes cluster that Radius is installed on is not compromised.
3. It is the responsibility of the Kubernetes cluster to authenticate users. Administrators and users with sufficient privileges can perform their required tasks. Radius cannot prevent actions taken by an administrator.

> 

Reviewer: Great 👍

   1. Communication:
      1. Controller and UCPD:
         1. Poll long-running operations (create, update, or delete) for a Recipe or a Deployment resource.
         2. Create a Radius Resource Group if needed.
         3. Create a Radius Application if needed.
         4. Get a Radius resource like the Environment that is associated with the resource.
         5. Create/Update/Delete a Recipe or a Deployment resource.
         6. Create/Update/Delete a Secret for a Recipe or a Deployment resource.
      2. Controller and the Kubernetes API Server:
         1. Fetch a Recipe or a Deployment object.
         2. Send events related to the operations running.
         3. Update a Recipe or a Deployment object.
         4. Create/Update/Delete a Secret associated with a Recipe or a Deployment object.
         5. List Deployments by filtering them by a specific Recipe object.

### Threats

> #### Threat: Users with access to the webhook server modifying the behavior of the webhook server

Reviewer: I think this is still unclear. As we discussed in the design meeting, we're assuming that the user has not tampered with Radius.

> If the user wants to crash their own car, that's allowed 😆

Reviewer: Let's sync up on this offline.

3. Ensure that all communication uses mTLS (Mutual TLS) to authenticate both the client and server, providing an additional layer of security. Verify that mTLS is correctly configured for all components and endpoints. Make changes to the necessary components if required.
4. Check if RBAC with Least Privilege is configured for every component to ensure that each component has only the permissions it needs to function. Make changes to the necessary components if required.
5. Define and implement necessary Network Policies to ensure that communication is accepted only from expected and authorized components. Regularly review and update these policies to maintain security.
6. Separate and firewall the etcd cluster to ensure the safety of the datastore. Implement network segmentation to isolate the etcd cluster from other components. Configure firewall rules to restrict access to the etcd cluster, allowing only authorized components and administrators to communicate with it. Regularly review and update firewall rules and network policies to maintain security.

> 7. Containers should run as a non-root user wherever possible to minimize the risks. Check if we can run any of the Radius containers as non-root. Do the necessary updates.

Reviewer: Let's revisit the action items after reviewing the threats.

# Title

- **Author**: ytimocin

## Overview

> This document aims to provide a threat model for the Radius Controller component. In this threat model, you will find potential security threats to the Controller component, an important part of Radius, and possible mitigations to the risks they pose. The document includes an analysis of the system, its assets, security objectives, identified threats, and recommended security measures to protect the system.

Reviewer: What's a `security objective`? The term is new to me.

Author: Removed this part.

> This document will not include designs for the proposed measures to each threat.

Reviewer: Why not? Are you saying that it won't included mitigations in the future? Or that it doesn't right now?

> 

Reviewer 2: Do we need a ## Terms and definitions section? eg. for ucpd

Author: Yes, we do. I will add that one too. Thanks for the reminder.

Reviewer 3: please add definition of UCPD and any other acronyms used in the doc.

## System Description

The Controller component is where we keep our Kubernetes controllers, a validating webhook, and a few other important parts. We will dive into more details on the controller below.

Note: If you would like to learn more about Kubernetes controllers, you can visit [this link](https://kubernetes.io/docs/concepts/architecture/controller/).

### Architecture

The Controller component consists of several key parts:

> - **Recipe and Deployment Controllers**: Custom Kubernetes controllers registered with the manager. Each controller is responsible for watching specific Kubernetes resources (Recipe and Radius Deployment) and reconciling their state.

Reviewer: Can you say more? I'm hoping the doc will go into some detail about what these controllers do.

Author: Added.

> - **Recipe Validating Webhook**: [Used if TLS certificates are provided] This webhook is triggered by the Kubernetes API in case of a change in a Recipe resource. Kubernetes API reaches out to this webhook in case of a creation, an update, or a deletion of a Recipe. Webhook validates the action, if it is a valid action, and responds to the Kubernetes API. For more information about webhooks, refer to the [official documentation](https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/).

Author: ```suggestion
- **Recipe Validating Webhook**: [Used if TLS certificates are provided] This webhook is triggered by the Kubernetes API in case of a change in a Recipe resource. Kubernetes API reaches out to this webhook in case of a creation, an update, or a deletion of a Recipe. Webhook validates the action and if it is a valid action, responds to the Kubernetes API. For more information about webhooks, refer to the [official documentation](https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/).
```

> - **Health Checks**: Health checks are implemented to monitor the status and performance of the controllers. They ensure that the controllers are functioning correctly and can take corrective actions if any issues are detected.

Reviewer: Hoping the doc will cover why this is relevant in the context of security.

### Clients

In this section we are going to be talking about different clients of the Controller component. Clients are systems that reach out to the Controller component to trigger an action. Here are the clients of the Controller component:

> 1. **Kubernetes API Server**: The primary client that interacts with the controller. It sends events related to resource changes (e.g., creation, update, or deletion of a Recipe and/or Deployment) to the controller manager. The controller watches for these events and reconciles the state of the resources accordingly.

Reviewer: Description seems somewhat confused between the API server "sending events" and the controller "watching for events".

> 1. **Validating Webhook Clients**: If webhooks are enabled and registered, clients that interact with the Kubernetes API server (e.g., `kubectl`, other controllers) can trigger webhook calls.

Reviewer: > If webhooks are enabled and registered

Is this something that users get to configure?

> clients that interact with the Kubernetes API server (e.g., `kubectl`, other controllers) can trigger webhook calls

Do these "other clients" interact with our webhook directly?

Author: I don't think any client other than Kubernetes API Server can talk to the Validating Webhook. Not sure if we need to specify this in the configuration of the Webhook: https://github.com/radius-project/radius/blob/main/deploy/Chart/templates/controller/validating-webhook-configuration.yaml#L28.

Based on what I have been reading for the last 20 minutes, Kubernetes API Server is the only client that can access a given Webhook. Correct me if I am wrong, @rynowak.

1. **Health Check Probes**: Kubernetes itself can act as a client by performing health and readiness checks on the controller manager.
1. **Metrics Scrapers**: If metrics are enabled, Prometheus or other monitoring tools can scrape metrics from the controller manager.

## Data Flow

### Diagram

> ![Controller Component via Microsoft Threat Modeling Tool](./2024-08-controller-component-threat-model/controller-component.png)

Reviewer 4: Would it make sense to show the user interaction, i.e., which trust boundaries are crossed by the user and which component(s) they interact with?

Author: I can add the User Interaction in this diagram too but in the Controller's case users don't directly reach out to the Controller. They make a change to a resource and then Kubernetes API Server reaches out to the Controller.

Author: `kubectl` is already authenticating and authorizing the user so it is not in Radius' scope. They have already crossed the trust boundary.

Author: Add user interaction to the diagram.

Author: Updated the diagram.

Reviewer 3: Should there be numbers on the arrows to indicate sequence of events? or is that in another place?

1. **Resource Creation/Update/Delete**: When a user creates, updates, or deletes a Recipe or a Radius Deployment resource, the request is sent to the Kubernetes API server.

> 2. **Webhook Validation/Mutation**: If a custom webhook is configured, the request is intercepted by the webhook for validation. The webhook ensures that the resource complies with the required policies. You can see the details of the Recipe Webhook implementation via [this link](https://github.com/radius-project/radius/blob/main/pkg/controller/reconciler/recipe_webhook.go#L46).

Reviewer: Can you describe what our webhook does? I think the link the code is helpful but not a substitute for the actual information.

Author: Updated accordingly.

3. **Controller Reconciliation**: When the action (create/update/delete) is validated by the Validating Webhook as discussed in the previous step, Kubernetes API reaches out to the Controller component and triggers the necessary action. Then, the controllers reconcile the resource's state to match the desired configuration.

> 4. **Communication with UCPD**: Controller also calls UCPD in some cases when doing reconcilations. The communication with UCPD is one-way, meaning that UCPD doesn't make calls to the Controller component, just responds to its requests. Controller reaches out to UCPD to read a resource and/or write to a resource. In some cases, it can request secrets for a given resource.

Reviewer: > Controller also calls UCPD in some cases when doing reconcilations.

Which cases?

Author: Added the cases below.

> ### Detailed Data Flows

Reviewer 4: I suggest moving this section after the others (Trust Boundaries, Assets and Security Objectives) because those sections provide additional context for why these particular risks are identified.

#### Kubernetes API Server - Validating Webhook for the Recipe Resource

> 

Reviewer: Suggestion: I think the formatting of this could be improved a lot. Right now each threat is a bullet point, which means either: 1) context is not provided, or 2) a lot of text needs to be on one line. 

Here's another idea.

```
##### Threat: A bad thing might happen

<description here>

**Mitigation**: <description here>

**Status:** <External|Active|Planned>

```

- **Assumptions**:

>   - TLS Certificates are defined on the startup of Controller component.

Reviewer: Let's talk about this assumption. It feels like we could say more.

Author: Removed this assumption because I see that TLS Certificates are always provided: https://github.com/radius-project/radius/blob/main/cmd/controller/main.go#L45.

- **Flow**:
  1. Kubernetes API Server receives a request to create, update, or delete a Recipe or Radius Deployment resource.

>   1. Kubernetes API Server sends the request to the Validating Webhook for validation.

Author: I need to double-check but validating webhook is just for the Recipe resource, I think.

  1. The webhook evaluates the request and responds with an approval or a rejection.

>   1. If approved, Kubernetes API Server proceeds with the action.

Reviewer: This description is correct overall but it's missing details about what our webhook does.

Author: Updated this section too.

> - **Risks and Mitigations**:

Reviewer: nit: be consistent about terminology. Risk -> Threat

>   - **Risk**: Could someone inject dangerous data into the resource object that would also get an approval from the webhook?

Reviewer: Do you have an example of what `dangerous data` would look like?

Is this threat specific to our scenario? Or is this a hypothetical for any webhook?

>     - **Mitigation**: Implement comprehensive validation logic in the webhook to check for malicious data patterns, enforce strict schema validation, and sanitize inputs.

Reviewer 5: > malicious data patterns

Were there specific patterns you were thinking of?

Author: Not sure if you can inject data through a resource definition in Kubernetes but that one was the main one that I was thinking of. I will have to investigate further to see if this could really pose a threat on our system.

>   - **Risk**: What if the webhook server is unavailable or slow to respond?

Reviewer: I'm glad you're covering this one. This is one of my favorites.

>     - **Mitigation**: Configure timeouts and retries for webhook calls, and implement fallback mechanisms to handle webhook failures gracefully.

Reviewer 5: Handling failure gracefully makes sense, but curious to know more about the potential risk unavailability or high latency poses.

    - **References**:
      - <https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#timeouts>

>       - <https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#failure-policy>

Reviewer: This sounds like a great topic for us to dig into more 👍

>   - **Risk**: What if the TLS certificates used for securing communication are compromised?

Reviewer: Can you say more? What does it mean for a certificate to be compromised?

Author: Going into more detail in another section.

    - **Mitigation**: Implement strong certificate management practices, including regular rotation of certificates, using short-lived certificates, and monitoring for certificate anomalies.

>   - **Risk**: Could an insider with access to the webhook server modify its behavior to approve malicious requests?

Reviewer 4: An additional mitigation could be to ensure that the Radius documentation has a clear guide for securing the deployment of Radius infrastructure.

>   - **Risk**: Could an insider with access to the webhook server modify its behavior to approve malicious requests?

Reviewer: I think you're saying something important here but it could be explained better. let's discuss.

    - **Mitigation**: Implement strict access controls, audit logs, and monitoring for changes to the webhook server configuration and code.
  - **Risk**: What if the webhook logic contains bugs that lead to incorrect validation decisions?

>     - **Mitigation**: Implement thorough testing, including unit tests, integration tests, and fuzz testing, to ensure the correctness of the webhook logic.

Reviewer 5: What is the worst case scenario if this does happen?

Author: The webhook will be returning approvals instead of rejections and the reconciler will be trying to reconcile objects that shouldn't be reconciled. I think we had a similar issue a few months ago. Controller was getting in a failed state. I will share the issue if I can find it.

> 

Reviewer: We expect the webhook to be invoked by the Kubernetes API server and only the Kubernetes API server. How is this secured?

Should we add any assumptions about the Kubernetes API server?

#### Kubernetes API Server - Radius Controllers - UCPD

- **Assumptions**:
  - Change to a Recipe or Radius Deployment resource are validated by the Validating Webhook.
- **Flow**:

>   1. The Kubernetes API Server communicates with the necessary controller to trigger the reconciler for the affected resource.

Reviewer: Is this true?

Author: Nope and removed this.

  1. The controller calls the function that will perform the reconciliation for the given resource.
  1. Based on the change, the controller can make calls to the UCPD to read details like resource information or secrets for the given resource, or to write the updated resource.
  1. UCPD responds to the request of the controller. The controller continues the process of reconciliation for the resource.
- **Risks and Mitigations**:

>   - **Risk**: Inconsistent state due to partial failures during reconciliation.

Reviewer: Is this a security issue?

Author: It is not. It is data consistency issue.

    - **Mitigation**: Implement idempotent operations in the controller to ensure that repeated executions do not cause inconsistent state, and use transaction mechanisms where applicable.
  - **Risk**: Unavailability of the UCPD.

>     - **Mitigation**: Implement high availability and redundancy for the UCPD, and use retry mechanisms with exponential backoff in the controller.

Reviewer 5: I'm not sure "implementing high availability" is the mitigation here, the risk assumes that it is never going to be 100%, so in the x% situations when it is unavailable, how do we handle it? Are retries acceptable? Because they can easily make the situation worse if the system is already overwhelmed. How do we gracefully reject requests?

Author: I removed this part since it is not really a security concern.

>   - **Risk**: Unauthorized access to the Kubernetes API Server.

Reviewer: Can you say more? What's an example of this?

Author: Talking more about this in another section.

    - **Mitigation**: Implement strong authentication and authorization mechanisms, such as RBAC (Role-Based Access Control) and network policies to restrict access.
  - **Risk**: Data tampering during communication between the controller and UCPD.

>     - **Mitigation**: Use TLS to encrypt communication between the controller and UCPD, and ensure proper certificate management.

Author: We are already using HTTPS by the way.

>     - **Mitigation**: Use TLS to encrypt communication between the controller and UCPD, and ensure proper certificate management.

Reviewer: Let's discuss 😆

>   - **Risk**: Controller logic errors leading to incorrect reconciliation.

Reviewer: Is this a security concern?

Author: No. Removed this one too.

>     - **Mitigation**: Implement thorough testing, including unit tests, integration tests, and end-to-end tests, to ensure the correctness of the controller logic.

Reviewer 5: Same question as above, what kind of issues/level of risk are we looking at if a bug gets introduced even with all the testing?

Author: Removed this part too.

>   - **Risk**: Performance bottlenecks due to high load on the controller.

Reviewer: Do we have evidence to believe this is likely? Or is this hypothetical?

>     - **Mitigation**: Optimize the controller code for performance, use efficient algorithms, and implement rate limiting to control the number of requests processed at a time.

Reviewer 5: I don't think the first two mitigations (Optimize the controller code for performance, use efficient algorithms) are really mitigations, since regardless of performance there will be some upper limit. So assuming the mitigation should be how the system handles load once the limit has reached.

Author: These are really not security related so I removed them.

  - **Risk**: Security vulnerabilities in the controller code.
    - **Mitigation**: Conduct regular security audits and code reviews, and follow secure coding practices to minimize vulnerabilities.

> ## Trust Boundaries

Author: I will definitely need to add more information here.

> ## Trust Boundaries

Reviewer: The discussion of trust boundaries should be placed earlier in the document. We've already made a list of threats/mitigations at this point.

We have a few different trust boundaries for the Controller component:

- Kubernetes cluster,
- Namespaces within the cluster.

The Controller component lives inside the `radius-system` namespace in the Kubernetes cluster where it is installed. UCPD also lives inside the same namespace.

The Kubernetes API Server, which is the main interactor of the Controller component, runs in the `kube-system` namespace within the cluster.

1. Different namespaces separate and isolate resources and workloads within the Kubernetes cluster.
2. Access controls and other permissions are implemented to manage interactions between namespaces.

> 

Reviewer: You're really close to landing these points here.

> ## Assets and Security Objectives

Reviewer: I'm not sure what assets or security objectives are. Can you explain?

Author: Removed this section.

### Assets

1. **Kubernetes Resources**

   - **Description**: Custom resources managed by the controllers: Recipes, and Deployments.
   - **Importance**: These resources are critical for the operation of the Kubernetes cluster and the applications running on it. Unauthorized access or modification could lead to service disruptions or security breaches.

1. **TLS Certificates**

   - **Description**: Certificates used for securing communication between the validating webhook and the Kubernetes API server.
   - **Importance**: Ensuring the confidentiality and integrity of communications. Compromised certificates could lead to man-in-the-middle attacks. [Related](https://stackoverflow.com/a/2109786).

1. **Configuration Files**

>    - **Description**: Configuration files that define the behavior of the controllers and the webhook. We have a configuration file for the Controller component and certificates for the Validating Webhook.

Author: Please see: [`controller-config.yaml`](https://github.com/radius-project/radius/blob/main/deploy/Chart/templates/controller/configmaps.yaml).

   - **Importance**: Incorrect or malicious configurations could lead to unintended behavior or security vulnerabilities.

> 1. **Logs and Monitoring Data**

Reviewer 4: Do we need a risk and mitigation for unauthorized access to logs?

Reviewer 4: Also, is there any risk of secrets leaking into the logs? Could that be a customer/user error that we could mitigate? Or something to add to the documentation to remind customers/users to not output secrets to logs.

   2. **Library**: The library used to calculate the hash of the secret is the crypto library, which is one of the standard libraries of Go: [Link to library](https://pkg.go.dev/crypto@go1.23.1).
   3. **Type**: [SHA1](https://www.rfc-editor.org/rfc/rfc3174.html). Note: "SHA-1 is cryptographically broken and should not be used for secure applications." [Link to warning](https://pkg.go.dev/crypto/sha1@go1.23.1). This is used as an optimization for detecting changes, not as a security protection.

#### Storage of secrets

Below you will find where and how Radius stores secrets. We create Kubernetes Secret objects and rely on Kubernetes security measures to protect these secrets.

1. **Deployment Reconciler**: Creates or updates a Kubernetes Secret for the connection values in a Deployment object. This Kubernetes Secret is deleted when the Deployment is deleted.
2. **Recipe Reconciler**: Creates a Kubernetes Secret for the Recipe object if it is defined in its spec. If there is an update to the secret object, the old one is deleted and the new one is added. When the Recipe is deleted, the Secret is also deleted.

#### Data Serialization / Formats

> We use custom parsers to parse Radius-related resource IDs and do not use any other custom parsers and instead rely on Kubernetes built-in parsers. Therefore, we trust Kubernetes security measures to handle data serialization and formats securely. The custom parser that parses Radius resource IDs has its own security mechanisms that don't accept anything other than a Radius resource ID.

Reviewer: We should have an action item to do fuzz-testing on the resource ID. 

Custom format + untrusted data -> do fuzz testing.

Authorized users with access to create a `Recipe` resource in Kubernetes can execute any Recipe in any Environment registered with Radius.

At the time of writing, Radius does not provide granular authorization controls. Any authenticated client can create any Radius resource and execute any action Radius is capable of taking. This is not limited to the Kubernetes controllers.

**Impact:** An authorized user of the Kubernetes cluster with permission to create a `Recipe` resource can execute any Recipe in any Environment registered with Radius.

**Mitigations:**

1. Operators should limit access to the `Recipe` resource using Kubernetes RBAC.
2. Operators should limit direct access to the Radius API using Kubernetes RBAC.
3. We should revisit the threat model and provide a more robust set of authorization controls when granular authorization policies are added to Radius.

**Status:** These mitigations are partial and require configuration by the operator. We will revisit and improve this area in the future.

## Open Questions

## Action Items

> 1. Use a hashing algorithm other than SHA-1 while computing the hash of the configuration of a Deployment object. This is a breaking change because deployments that are already hashed with SHA1 should be redeployed so that reconciler can work as expected.

Reviewer: This isn't a breaking change, we'd just do extra work 😆

2. Check if RBAC with Least Privilege is configured for every component to ensure that each component has only the permissions it needs to function. Make changes to the necessary components if required.
3. Define and implement necessary Network Policies to ensure that communication is accepted only from expected and authorized components. Regularly review and update these policies to maintain security.

> 4. Containers should run as a non-root user wherever possible to minimize the risks. Check if we can run any of the Radius containers as non-root. Do the necessary updates.

Reviewer: I'd prefer that these "review" items be handled as part of the threat model review.

For example:

> Check if RBAC with Least Privilege is configured for every component to ensure that each component has only the permissions it needs to function.

The threat model is the **right place for us to define** the "permissions each component needs to function". It doesn't make sense to push this work out of the threat model. 

---

> Containers should run as a non-root user wherever possible to minimize the risks. Check if we can run any of the Radius containers as non-root.

This is just scratching the surface in terms of policies applied to our components 😆. See: https://kubernetes.io/docs/concepts/security/pod-security-standards/#policy-instantiation

> # Title

Reviewer: Can you update the title?

- **Author**: ytimocin

## Overview

This document provides a threat model for the Radius Controller component. It identifies potential security threats to this critical part of Radius and suggests possible mitigations. The document includes an analysis of the system, its assets, identified threats, and recommended security measures to protect the system.

The Radius Controller component monitors changes (create, update, delete) in the definitions of Recipe and Deployment resources. Based on these changes, the appropriate controller takes the necessary actions. Below, you will find detailed information about the key parts of the Radius controllers and the objects they manage.

> ## Terms and Definitions

Author: I will be adding more terms and definitions here.

| Term                  | Definition                                                                                                                                                                                  |
| --------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| Admission Controllers | An admission controller is a piece of code that intercepts requests to the Kubernetes API server prior to persistence of the object, but after the request is authenticated and authorized. |
| mTLS                  | Mutual Transport Layer Security (mTLS) allows two parties to authenticate each other during the initial connection of an SSL/TLS handshake.                                                 |

> ## System Description

Reviewer 6: I have a meta-question - when we change these components, this doc will be out of date. Is the expectation that we update these on feature implementations or do we just write a new one every once in a while?

Author: Good question. We can talk about this. I believe that whenever there is an important update, we need to go and update this document too.

For example, if we update the logic of the existing validating webhook or if we add a mutating webhook, we should update the threat model too.

Note: Kubernetes has [admission controllers](https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/) that intercept requests to the Kubernetes API Server before the persistence of the object. Admission controllers may be **validating**, **mutating**, or both. Here is a simple diagram of the flow from the command entered by the user to the persistence of the object to etcd.

![Admission Controllers' Flow](./2024-08-controller-component-threat-model/admission-controllers-flow.png)

### Architecture

The Controller component consists of several key parts:

- **Recipe and Deployment Controllers**: Kubernetes controllers registered with the manager. Each controller is responsible for watching specific Kubernetes resources (Recipe and Deployment in our case) and reconciling their state.
  - **Recipe Controller**: This controller specifically watches for additions of and changes in Recipe objects in the cluster. The controller calls UCPD to perform the following operations:
    - Create, update, or delete a Recipe object based on the input received from the Kubernetes API Server.

>     - Create, update, or delete a Secret if it is requested by the Recipe object.

Reviewer: These two bullet points are vague. Can we be more specific? Maybe it would be helpful to start with the purpose of this controller and then describe what the outputs are.

  - **Deployment Controller**: This controller specifically watches for additions of and changes in Deployment objects in the cluster. The controller calls UCPD to perform the following operations:
    - Create, update, or delete a Deployment object based on the input received from the Kubernetes API Server.

>     - Create, update, or delete a Secret if it is requested by the Deployment object.

Reviewer: Same comments as the Recipe controller, this is vague and not really accurate.

1. **Health Check Probes**: Kubernetes itself can act as a client by performing health and readiness checks on the controller manager.
1. **Metrics Scrapers**: If metrics are enabled, Prometheus or other monitoring tools can scrape metrics from the controller manager.

## Trust Boundaries

We have a few different trust boundaries for the Controller component:

- **Kubernetes Cluster**: The overall environment where the Controller component operates and receives requests from the clients.
- **Namespaces within the Cluster**: Logical partitions within the cluster to separate and isolate resources and workloads.

The Controller component lives inside the `radius-system` namespace in the Kubernetes cluster where it is installed. UCPD also resides within the same namespace.

The Kubernetes API Server, which is the main interactor of the Controller component, runs in the `kube-system` namespace within the cluster.

> 

Reviewer: What are the security implications of namespaces? Can you elaborate on why this is relevant information in the context of security?

Author: Added below in the **Key Points of Namespaces** section.

### Key Points

> 1. **Namespace Isolation**: Different namespaces separate and isolate resources and workloads within the Kubernetes cluster.

Reviewer: Isolate how?

2. **Access Controls**: Access controls and other permissions are implemented to manage interactions between namespaces.

## Data Flow

### Diagram

![Controller Component via Microsoft Threat Modeling Tool](./2024-08-controller-component-threat-model/controller-component.png)

1. **Resource Creation/Update/Delete**: When a user requests to create, update, or delete a Recipe or Deployment object, the request is sent to the Kubernetes API Server. For example, this request can be made by a human interactor using the `kubectl` command. The Kubernetes API Server authenticates and authorizes the request before the admission controllers intercept it.

2. **Validating Webhook**: The only type of admission controller we have in Radius is the validating webhook for the Recipe resource. The validating webhook ensures that the Recipe object is one of the Radius portable resources. If the operation and the object is validated, then it is persisted to the **etcd**. You can see the details of the Recipe Webhook implementation via [this link](https://github.com/radius-project/radius/blob/main/pkg/controller/reconciler/recipe_webhook.go#L46).

> 3. **Controller Reconciliation**: When the action (create/update/delete) is validated by the Validating Webhook (applicable only to the Recipe resource and not to the Deployment resource) as discussed in the previous step, the Kubernetes controllers we have (Recipe and Deployment Reconcilers) watch for changes via the Kubernetes API Server. The controllers then reconcile the resource's state to match the desired configuration specified in the resource definition. This involves ensuring that the actual state of the cluster matches the desired state, making any necessary adjustments to achieve this.

Reviewer 3: is the reconciliation mechanism something we implemented or is owned by kubernetes?

Author: The Recipe and Deployment reconcilers are implemented by us.

4. **Communication with UCPD**: Controller also calls UCPD when doing reconcilations. Here is the list of instances where Controller calls UCPD:

   1. Recipe Reconciler:

      - Poll operations like create, update, or delete a Recipe resource.
      - Resolve dependencies for a Recipe resource:
        - Get the Radius environment the Recipe is associated with.
        - Create the Radius Resource Group if needed.
        - Create the Radius Application if needed.
      - Fetch a Radius resource.
      - List Secrets of a Recipe resource.

>       - Delete a Recipe resource.

Reviewer: Should we give an example? I think this is technically accurate but still hard to grasp.

   2. Deployment Reconciler:

      - Poll operations like create, update, or delete a Deployment resource.
      - Resolve dependencies for a Deployment resource:
        - Get the Radius environment the Deployment is associated with.
        - Create the Radius Resource Group if needed.
        - Create the Radius Application if needed.
      - Fetch a Recipe resource.

>       - List Secrets of a Recipe resource.

Reviewer: What else?

5. **Communication with Kubernetes API Server**: Controller also calls the Kubernetes API Server when doing reconcilations. Here is the list of instances where Controller calls the Kubernetes API Server:

   1. Recipe Reconciler:

      - Fetch a Recipe object.
      - Send events related to the operations.
      - Update a Recipe object:
        - Status update.
        - Addition and removal of a Finalizer.
      - Delete a Secret object that is associated with the Recipe.
      - Create a Secret object to associate with the Recipe if needed.

   2. Deployment Reconciler:

      - Fetch a Deployment object.
      - Send events related to the operations.
      - Update a Deployment object:
        - Status update.
        - Addition and removal of a Finalizer.
      - Delete a Secret object associated with a Deployment that is no longer needed.
      - Fetch a Secret object.
      - Create or Update a Secret object and associate it with a Deployment object.
      - List Deployments filtered with specific Recipe objects.

### Threats

> #### Impersonation of Radius Controllers to trigger resource access

Author: Access to the Radius API:
- everybody (admin users) can send requests to the Radius API

What are all the possible clients of Radius API? We own and other users own.

RBAC can solve this issue instead of Network Policies

mTLS, Network Policies, RBAC

Risk: An attacker can impersonate the Radius Controllers to talk to UCPD.

Impact: The Controller Component and UCPD communicate in order to create/update/delete necessary resources and these resources can also be cloud resources. This can lead to loss of resources or creation/update of existing resources.

Mitigation:

1. **mTLS**: mTLS should be (and is) enabled for communication between the Controller and the UCPD.

> 2. **Network Policies**: We can add necessary policies that UCPD can only accept communication from the Controller and other important components that it is should be talking to.

Reviewer 7: Can you be specific? My concern is that depending on the way this is implemented there could be risks that allow bypassing of some network strategies.

Author: This is an additional policy that will let UCP only talk to the allowed pods/components in the allowed namespaces.

> 2. **Network Policies**: We can add necessary policies that UCPD can only accept communication from the Controller and other important components that it is should be talking to.

Reviewer: Is this reasonable? What about other clients?

> 2. **Network Policies**: We can add necessary policies that UCPD can only accept communication from the Controller and other important components that it is should be talking to.

Reviewer 3: we should or we will?  does this belong in "best practices" or will we enable/enforce it?

Author: I added an action item to add these policies to all the necessary components.

> #### Compromised TLS Certificates and private keys of the Validating Webhook

Author: You need to register as the validating webhook so that Kube API Server can send you the requests.

If it is admin doing something, then it is admin, and they can do what they want.

If not admin, then can they act as the Validating Webhook?

Risk: If the TLS certificates and the private keys used by the validating webhook are compromised, an attacker could impersonate the webhook server, intercepting the requests from the Kubernetes API Server or altering the responses.

> Impact: The validating webhook functions should return an approval or a validation error to the caller. The impersonater can return incorrect responses to the Kubernetes API Server which could affect the lifecycle of resources.

Reviewer 6: ```suggestion
Impact: The validating webhook functions should return an approval or a validation error to the caller. The impersonator can return incorrect responses to the Kubernetes API Server which could affect the lifecycle of resources.
```

Risk: If the TLS certificates and the private keys used by the validating webhook are compromised, an attacker could enable unauthorized clients to communicate with the webhook.

> Impact: Unauthorized clients wouldn't have any affect in the lifecycle of resources because, even if the webhook returns an approval or a rejection to the unauthorized client, the unauthorized client should also have access to the Kubernetes API Server to further trigger changes in the resource.

Reviewer 3: I don't understand this impact. Can we discuss?

Mitigation:

1. **mTLS (Mutual TLS)**: Implement mutual TLS to ensure that both the client (Kubernetes API Server) and the server (Validating Webhook) authenticate each other. This helps in verifying the authenticity of both parties and prevents unauthorized access. Note that mTLS is enabled for the communication between the Validating Webhook and the Kubernetes API Server in Radius as of now.
2. **Network Policies**: Implement necessary Network Policies to ensure that communication between components is restricted to authorized components only. In this case, the Validating Webhook and the Kubernetes API Server should only accept requests from each other.
3. **Strong Certificate Management**: Implement strong certificate management practices, including regular rotation of certificates, using short-lived certificates, and monitoring for certificate anomalies.

> 4. **Secure Storage**: Store TLS certificates securely using secrets management solutions to prevent unauthorized access. The certificates and the private keys are being kept in a Secret object on Kubernetes.

Reviewer: Which of these mitigations are in place already and which require us to do work? Which require the user to do work?

#### Compromised TLS Certificates and private keys of the Kubernetes API Server

Risk: If the TLS certificates and private keys used by the Kubernetes API Server are compromised, an attacker could impersonate the Kubernetes API Server. This would allow the attacker to intercept, alter, or forge communication between the API Server and other components, including the Validating Webhook and Reconcilers. This could lead to unauthorized access, data breaches, and manipulation of cluster operations.

Impact:

1. The attacker could gain unauthorized access to sensitive data including the data it can get from the etcd if proper policies are not in place.

> 2. The attacker can manipulate the operations of the cluster, causing unauthorized changes.

Reviewer: Are these threats related to our components?

Author: Actually, no. I am removing the threat that includes the compromise of certs and private key of Kubernetes API Server.

Mitigation:

1. **Network Policies**: Implement necessary Network Policies to ensure that communication between components is restricted to authorized components only. For example, the Validating Webhook should only accept requests from the original Kubernetes API Server.
2. **Strong Certificate Management**: Implement strong certificate management practices, including regular rotation of certificates, using short-lived certificates, and monitoring for certificate anomalies.
3. **Secure Storage**: Store TLS certificates securely using secrets management solutions to prevent unauthorized access. The certificates and the private keys are being kept in a Secret object on Kubernetes.

#### Interception and manipulation of the communication between the webhook and the Kubernetes API Server

**Risk:**

> 1. An attacker could intercept and manipulate the communication between the webhook and the Kubernetes API Server. This could lead to unauthorized access, data breaches, or manipulation of admission decisions, potentially compromising the security and integrity of the Kubernetes cluster.

Reviewer: It feels like this is getting repetitive. Three of the threats listed so far are attacking authentication between the API Server and the webhook. 

We're missing an explanation of how this is secured today.

**Mitigation:**

1. **mTLS (Mutual TLS)**: Implement mutual TLS to ensure that both the client (Kubernetes API Server) and the server (Validating Webhook) authenticate each other. This helps in verifying the authenticity of both parties and prevents unauthorized access. Note that mTLS is enabled for the communication between the Validating Webhook and the Kubernetes API Server in Radius as of now.
2. **Network Policies**: Use Kubernetes Network Policies to restrict the network traffic between the Kubernetes API Server and the webhook. This limits the exposure of the webhook to only trusted sources. Reference: <https://kubernetes.io/docs/concepts/services-networking/network-policies/>.

> #### An attacker gaining access to etcd and manually changing resources

Reviewer 3: Is this our vulnerability or Kubernetes? i.e. how would Radius either make it easier (or harder) to access etcd?

Author: Actually, yes, this is a Kubernetes-related security issue and not related to the Controller component or Radius. Removing this threat.

Risk: If an attacker gains access to etcd, they can manually change resources stored in the etcd datastore. This could lead to unauthorized access, data breaches, and manipulation of cluster operations.

Mitigation:

> 1. **Access Controls for etcd**: Implement strict access controls for etcd to ensure that only authorized components can access and modify its data.

Reviewer 7: What access controls?

2. **Network and Firewall Policies**: Add the necessary Network and Firewall Policies to filter access to the etcd.

> 3. **Audit Logs**

Reviewer 7: Would this be an appropriate mitigation for the other risks as well? anomaly identification could be applied to any audit stream.

4. **Encryption**

> 

Reviewer: This is a good example of something that's out of scope 👍 You'll notice it doesn't talk about any of our components. I think we should capture this as an assumption somewhere. Let's discuss in the meeting.

> 

Reviewer 8: Would we also dive into the exact security configurations as part of this effort?

#### Dangerous data hidden in the Recipe or Deployment resource

> Risk: When an external interactor sends a create or update request for a Recipe or Deployment, there is a risk that the interactor could hide dangerous data within the resource. This hidden data could potentially exploit vulnerabilities in the system, leading to security breaches or system malfunctions.

Reviewer 7: I'm unclear what the risk is today, all the mitigations speak about custom parsers that are not currently in use.

Author: Yes, today we don't have this risk because we don't have a custom parser, but I just wanted to talk about this too just in case if in future we would like to add a custom parser or use a third party one.

Author: On a second thought, I will remove this since this is not a threat as of now. If we introduce custom parsers, then we can add a section on that.

Mitigation:

1. **Use of Kubernetes Parser**: Radius relies on the parser provided by Kubernetes, which is well-tested and maintained. This reduces the risk associated with parsing malicious data.
2. **Avoid Custom Parsers**: By not using custom or third-party parsers, Radius avoids the additional risk and complexity associated with ensuring the security of these parsers.

> 3. **Fuzz Testing**: If a custom parser were to be used, it would be essential to implement fuzz testing. Fuzz testing involves providing invalid, unexpected, or random data inputs to the parser to identify potential vulnerabilities. For more information on fuzz testing, refer to [OWASP Fuzzing](https://owasp.org/www-community/Fuzzing).

Reviewer: This one is good. Can we get more specific? Do we have any custom parsers?

Author: I thought about removing this since we don't have a custom parser in place as of now. That is why, I removed this threat. We can talk about this when we actually have a custom parser or use a 3rd party one. What do you think?

#### Users with access to the webhook server modifying the behavior of the webhook server

Risk: Could an insider with access to the webhook server modify its behavior to approve malicious requests?

> Mitigation: Implement strict access controls, audit logs, and monitoring for changes to the webhook server configuration and code.

Reviewer: This is a great one to discuss. It falls in the category of "what is a valid use case" and what is not.

#### Unauthorized access to the Kubernetes API Server

> 

Reviewer 3: as above, is this a Radius vulnerability or Kubernetes only?

Author: Not a Radius security issue. Removing this threat too.

Author: Thanks!

- **Mitigation**: Implement strong authentication and authorization mechanisms, such as RBAC (Role-Based Access Control) and network policies to restrict access.

#### Security vulnerabilities in the controller code

> - **Mitigation**: Conduct regular security audits and code reviews, and follow secure coding practices to minimize vulnerabilities.

Reviewer 7: Is there an example from another open source project to point to?

## Open Questions

<!--
List any unresolved questions or uncertainties about the threat model. Use this section to gather feedback from experts or team members and to track decisions made during the review process.
-->

## Action Items

1. Check if TLS is enabled for every component to ensure secure communication. Make changes to the necessary components if required.
2. Ensure that all communication uses mTLS (Mutual TLS) to authenticate both the client and server, providing an additional layer of security. Verify that mTLS is correctly configured for all components and endpoints. Make changes to the necessary components if required.

> 3. Check if RBAC with Least Privilege is configured for every component to ensure that each component has only the permissions it needs to function. Make changes to the necessary components if required.

Reviewer: Can **you** check this?

> 4. Define and implement necessary Network Policies to ensure that communication is accepted only from expected and authorized components. Regularly review and update these policies to maintain security.

Reviewer: What do network policies do? Is there any configuration the user has to provide?

Hint: https://github.com/argoproj/argo-cd/security/advisories/GHSA-9766-5277-j5hr

5. Separate and firewall the etcd cluster to ensure the safety of the datastore. Implement network segmentation to isolate the etcd cluster from other components. Configure firewall rules to restrict access to the etcd cluster, allowing only authorized components and administrators to communicate with it. Regularly review and update firewall rules and network policies to maintain security.

> 6. Containers should run as a non-root user wherever possible to minimize the risks. Check if we can run any of the Radius containers as non-root. Do the necessary updates.

Reviewer: Can **you** check this?

> ## Review Notes

Reviewer 8: What if user registered a malicious recipe or just versions of resource that have known vulnerabilities? the audit logs could capture who/ when of the event, but would the scenario fall under radius related security?

## Terms and Definitions

| Term                  | Definition                                                                                                                                                                                  |
| --------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| Admission Controllers | An admission controller is a piece of code that intercepts requests to the Kubernetes API server prior to persistence of the object, but after the request is authenticated and authorized. |
| mTLS                  | Mutual Transport Layer Security (mTLS) allows two parties to authenticate each other during the initial connection of an SSL/TLS handshake.                                                 |
| UCPD                  | Universal Control Plane Daemon for Radius                                                                                                                                                   |

## System Description

> The Controller component is a critical part of the Radius system, responsible for managing Kubernetes resources through custom controllers and webhooks. It ensures that the desired state of the system is maintained by continuously monitoring and reconciling resources.

Reviewer: ```suggestion
The Controller component enables users to manage Radius resources through the Kubernetes API. Users define their application using existing Kubernetes resource types like `Deployment` and the `Recipe` CRD provided by Radius. The job of the controller is to ensure that the desired state of the system is maintained in both Radius and Kubernetes by continuously monitoring and reconciling resources.
```

Suggestion to be more explicit about what the controller actually is for.

1. **Metrics Scrapers**: If metrics are enabled, Prometheus or other monitoring tools can scrape metrics from the controller manager.

## Trust Boundaries

We have a few different trust boundaries for the Controller component:

- **Kubernetes Cluster**: The overall environment where the Controller component operates and receives requests from the clients.
- **Namespaces within the Cluster**: Logical partitions within the cluster to separate and isolate resources and workloads.

The Controller component lives inside the `radius-system` namespace in the Kubernetes cluster where it is installed. UCPD also resides within the same namespace.

The Kubernetes API Server, which is the main interactor of the Controller component, runs in the `kube-system` namespace within the cluster.

### Key Points of Namespaces

1. **Isolation of Resources and Workloads**: Different namespaces separate and isolate resources and workloads within the Kubernetes cluster.
2. **Access Controls and Permissions**: Access controls and other permissions are implemented to manage interactions between namespaces.

> 3. **Separation of Concerns**: Namespaces support the separation of concerns by allowing different teams or applications to manage their resources independently, reducing the risk of configuration errors and unauthorized changes.

Reviewer: This is much better, thank you. I think we could still be more specific about what "isolation" means. Isolation how?

I think this is lower urgency than the threats and mitigiations.

## Data Flow

### Diagram

![Controller Component via Microsoft Threat Modeling Tool](./2024-08-controller-component-threat-model/controller-component.png)

1. **User creates/updates/deletes a Recipe or a Deployment resource**: When a user requests to create, update, or delete a Recipe or a Deployment resource, the request is handled by the Kubernetes API Server. One way, and probably the most common way, a user can do this request is by running a **kubectl** command. Kubernetes takes care of the authentication and the authorization of the user and its request(s) so we (Radius) don't need to worry about anything here.

2. **Validating Webhook**: The only type of admission controller we have in Radius is the validating webhook for the Recipe resource. The validating webhook ensures that the Recipe object to be created or updated is one of the Radius portable resources. Whenever Kubernetes API Server receives a request to create or update a Recipe object, it communicates the proposed changes with the validating webhook. If the validating webhook validates the changes, then it is persisted to the **etcd** by the Kubernetes API Server.

>    1. Possible threats:

Reviewer 5: Curious why we did we separate out threats in two sections (this and the section after it)?

      1. A bad actor pretending to be the Kubernetes API Server and communicating with the Validating Webhook. This is not a real threat since validating webhook only returns an approval or a rejection to the proposed changes. It is not doing any updates to the resources. That is the reason there is no real threat here.

>       2. A bad actor pretending to be the Validating Webhook and receiving the requests from the Kubernetes API Server.

Reviewer: I'd suggest combining this with the Threats section of the document.

         5. Create/Update/Delete a Recipe or a Deployment resource.
         6. Create/Update/Delete a Secret for a Recipe or a Deployment resource.
      2. Controller and the Kubernetes API Server:
         1. Fetch a Recipe or a Deployment object.
         2. Send events related to the operations running.
         3. Update a Recipe or a Deployment object.
         4. Create/Update/Delete a Secret associated with a Recipe or a Deployment object.
         5. List Deployments by filtering them by a specific Recipe object.
   2. Possible threats:
      1. The controller component should have the minimum necessary Role-Based Access Control permissions to perform its tasks. It shouldn't have the permissions that it is not supposed to use.
      2. A bad actor pretending to be UCPD, in case of compromised certificates, may lead to inconsistent resource data. The question is if a bad actor can create a process in the `radius-system` namespace that is going to pretend to be UCPD.
      3. In case of compromised certificates of the Controller component, a bad actor can pretend to be the Controller and make requests to the UCPD and Kubernetes API Server.

> 

Reviewer: Are there any assumptions we should make to simplify the analysis?

Author: Added them.

### Threats

#### Threat: A bad actor pretending to be the Controller component to communicate with UCPD

> In case of compromised certificates and private key, an attacker can pretend to be the Controller component to communicate with UCPD.

Reviewer: We should talk about the state of authentication within Radius. I think this is the right idea but some of the details are wrong.

**Impact**: Controller Component and UCPD communicate in order to create/update/delete necessary resources. The impact of this threat would be unwanted operations on the resources.

> **Mitigation**:

Reviewer 5: Have you considered including detection in the template as well? It helps in understanding we have detection mechanisms in place if it gets introduced after initial mitigations are implemented.

> 1. **mTLS**:

Author: We don't have mTLS. Action item go check.

   - **Description**: Ensure that mutual TLS is enabled for communication between the Controller and UCPD. This ensures that both parties authenticate each other, preventing unauthorized entities from impersonating the Controller.

>    - **Status**: Active

Reviewer 5: What does "status" indicate?

Author: Status indicates if the mitigation is already in place or not.

2. **Strong Certificate Management**:

   - **Description**: Implement strong certificate management practices, including regular rotation of certificates, using short-lived certificates, and monitoring for certificate anomalies. This reduces the risk of certificate compromise.
   - **Status**:

> 3. **Secure Storage**:

Reviewer 5: For the mitigations that are already implemented, can we be more specific about how it is implemented? For example instead of "secrets management solutions", we should specify the specific secret manager used.

   - **Description**: Store TLS certificates and private keys securely using secrets management solutions to prevent unauthorized access. Ensure that these secrets are encrypted and access is restricted to authorized components only.
   - **Status**:

4. **RBAC (Role-Based Access Control)**:

>    - **Description**: Implement strict RBAC policies to ensure that only authorized components and users have the necessary permissions to access and manage the Controller and UCPD. This minimizes the risk of unauthorized access.

Reviewer 5: Is this in addition to what's mentioned in the previous section about relying on Kubernetes auth? 

`Kubernetes takes care of the authentication and the authorization of the user and its request(s) so we (Radius) don't need to worry about anything here.`

Author: This one is removed, and you can see the new threats in the threat list.

   - **Status**:

3. **Secure Storage**:

   - **Description**: Store TLS certificates and private keys securely using secrets management solutions to prevent unauthorized access. Ensure that these secrets are encrypted and access is restricted to authorized components only.
   - **Status**:

4. **RBAC (Role-Based Access Control)**:

   - **Description**: Implement strict RBAC policies to ensure that only authorized components and users have the necessary permissions to accessthe Validating Webhook (in this case; Kubernetes API Server). This minimizes the risk of unauthorized access.
   - **Status**:

5. **Audit Logs**:

   - **Description**: Enable detailed audit logging for all interactions between the Validating Webhook and the Kubernetes API Server. Regularly review these logs to detect any unauthorized or suspicious activities.
   - **Status**:

6. **Network Policies**:

   - **Description**: Implement necessary network policies to ensure that communication only from the Kubernetes API Server is accepted by the Validating Webhook.
   - **Status**:

#### Threat: Users with access to the webhook server modifying the behavior of the webhook server

> A user with access to the webhook server can modify its behavior to approve malicious requests, which can be reconciled in the next step of the data flow. Reconciliation loops can also trigger resource updates or even deletions by calling UCPD with the malicious requests.

Reviewer: This is a great one for us to discuss in the context of assumptions.

Author: Assumptions: the user hasn't tampered with the Radius installation, Kubernetes cluster is not compromised.

**Impact**:

1. **Unauthorized Operations**: With the approval of malicious requests, unauthorized operations could be performed on user resources, including the creation, update, or deletion of resources.
2. **Resource Deletion**: Malicious modifications could lead to the deletion of existing resources, causing potential data loss and service disruption.

**Mitigation**:

1. **Audit Logs**:

   - **Description**: Implement detailed audit logging to track which user performs which operation on the webhook server. Regularly review these logs to detect any unauthorized or suspicious activities.
   - **Status**:

2. **RBAC (Role-Based Access Control)**:

   - **Description**: Implement strict RBAC policies to ensure that only authorized users have the necessary permissions to access and modify the webhook server. This minimizes the risk of unauthorized access and modifications.
   - **Status**:

#### Threat: Security vulnerabilities in the controller code

> Security vulnerabilities in the controller code can be exploited by attackers to gain unauthorized access, escalate privileges, or disrupt the normal operations of the Kubernetes cluster. These vulnerabilities can arise from things like coding errors, insecure coding practices, or outdated dependencies.

Reviewer: Can we be more specific? It's the goal of the threat model to identify specific risks not general risks.

2. **Privilege Escalation**: Attackers can escalate their privileges within the cluster, gaining access to sensitive resources and data.
3. **Data Breaches**: Sensitive information can be exposed, leading to data breaches.

**Mitigation**:

1. **Code Reviews**:

   - **Description**: Implement a thorough code review process to ensure that code changes are reviewed and approved by multiple team members.
   - **Status**:

2. **Dependency Management**:

   - **Description**: Regularly update and manage dependencies to ensure that the controller code is using the latest, most secure versions of libraries and frameworks. We can also use tools like dependabot to make sure to update to the most recent and secure version of the dependencies we are using.
   - **Status**: Active

3. **Static Analysis Tools**:

   - **Description**: Use static analysis tools to automatically scan the code for potential security vulnerabilities and coding errors. These tools can be added to the CI/CD pipelines that we already have in place.
   - **Status**: Active/Planned

#### Threat: Webhook server being unavailable or slow to respond

> If the webhook server becomes unavailable or slow to respond, it can lead to delays or failures in processing requests. This may not be a direct security issue but it can affect the overall reliability of the system.

Reviewer: nit: Availability is part of security. If malicious actors can cause downtime that's definitely relevant.

Is there a specific risk or vulnerability being identified here?

Author: What are some specific things that could cause this?

Author: Add UCP
